### ISSUE TITLE
Memory overconsumption + slowness for coverage over bed defined regions 

### ISSUE THREAD

**Issue by Sebastien-Raguideau** [brentp/mosdepth#71]
Hello,
So I'm trying mosdepth and having issue over memory consumption. I get over consumption of memory with around 600G of ram taken to deal with a 6G bam file. Though I need coverage for over 18 millions regions, which may be related to the issue. 
I am launching the following command : 
mosdepth -t 150 -n  -b Annotation/Bed_file.bed test/Sample_name Map/Sample_name.mapped.sorted.bam > test/log
For some reason execution is somehow slow. When i run a minimal test with 1000 regions in a bed file, it takes : 15 mins. Trying bedtool coverage, I get similar coverage but faster : 7 mins. 
I use mosdepth from a conda installation in a conda environment. 
So, would you have any insight on why this is happening? 
Thanks,
Seb

**Comment by brentp** [comment]
I will try some tests with a very large regions file. I don't know that this would cause the issue, but using `-t 150` is not helpful. The threads are decompression threads so the maximum improvement will come at about `-t 4`. Could you try that and verify you still see the issue?

**Comment by brentp** [comment]
I am able to run on a BED file with 22 million entries. Could you try this:
```
for chr in $(seq 1 22); do seq 1 1000000 | awk -vchr=$chr '{ print chr"\t"$0"\t"$0+10000 }'  ; done > o.bed
```
(assuming your chroms are 1..22, if not you'll have to adjust the for loop.)
and then use that as the -b argument to mosdepth? If that succeeds relatively quickly, then we can know it's something about your bed file. If not, it's something about your bam file.

**Comment by Sebastien-Raguideau** [comment]
Hello again, 
I tried the same as before but this time with only 10 contigs instead of 1000. 
bedtool took 6m58.455s while mosdepth took only 0m51.000s.
This is much better, but it looks like to me that the total number of contigs does impact significantly mosdepth speed, and memory.  
Best,
Seb

**Comment by brentp** [comment]
would you try the attached binary (after unzipping) and see if it suffers the same problems? It is a small change that might make the difference.

[mosdepth_ctgs.gz](https://github.com/brentp/mosdepth/files/2817753/mosdepth_ctgs.gz)

**Comment by Sebastien-Raguideau** [comment]
I can't make it work, it says that libhts.so is missing, I found other issue with libhts.so where you advise to export libhts.so before running mosdepth, I do, it is not working. Also, you advise to run  ldd $LD_LIBRARY_PATH/libhts.so, not sure what it does, but here is the output :
	linux-vdso.so.1 =>  (0x00007ffdaa1bd000)
	libdeflate.so => not found
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f2c8a0f1000)
	libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f2c89ee1000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f2c89cc7000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f2c899be000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f2c897ba000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f2c8959d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2c891d3000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f2c8a5d7000)

**Comment by brentp** [comment]
ah. ok. I built on my machine and that usually works for most people. For the release build, I use an older libc inside a docerk container. let me see if I can get that for this one.

**Comment by Sebastien-Raguideau** [comment]
Just tested it, I get the same issue, it ask for libhts.so.
I exported the path to it in  LD_LIBRARY_PATH, just before running mosdepth_fix2 and I get : 
could not load: libhts.so
compile with -d:nimDebugDlOpen for more information

**Comment by brentp** [comment]
hmm. this was built with the same method as the one from releases. The old mosdepth binary works in the same environment?

**Comment by johnomics** [comment]
Sorry to say I have a similar problem, 78k contigs, 1.6 Gb genome, attempting to calculate coverage for 1kb windows, mosdepth 0.2.4 using all RAM on any machine then crashing (tested on 768 GB RAM machine). Can't run the fixed binary either, get the same libhts.so error, also after installing mosdepth through conda. Any help greatly appreciated.

**Comment by brentp** [comment]
I'd like to fix this, but I haven't been able to recreate. Can you get me a full test-case of bam + bed that reproduces? I imagine it's relatively independent of the bam so you can do e.g.:
```
samtools view -s 0.01 -T $reference -C -o sub.cram $bam
```
to get me a sub-sampled cram file. Then send the reference, sub.cram and the bed and I can debug.

BTW, the output of `ldd` above shows that libdeflate is missing so that's not a problem I can resolve, you'll need a working libhts.so + environment.

**Comment by brentp** [comment]
I got a bam from a colleague with quite a large number of contigs. mosdepth is somewhat slow for that bam even without a bed file. I'll have a look at improving using that as a test-case and hope that will improve the use-case described in this issue.

**Comment by brentp** [comment]
hi, would one of you verify this observation?  when running with bam, there is the reported problem with memory use. when running with cram, there is no such problem.

**Comment by brentp** [comment]
also, could you verify if the -t (--threads) option is not used, there is no problem with memory?

**Comment by brentp** [comment]
I have posted an issue in htslib for this: https://github.com/samtools/htslib/issues/822

**Comment by johnomics** [comment]
Thank you very much for looking into this and raising the htslib issue. I can confirm that leaving off the -t option appears to resolve the problem. I have had a job running for 7 minutes, using ~15 GB of RAM. Yesterday the same job with the -t option used over 600 GB in a similar time.

Sorry I haven't had time to subsample the BAM or convert to CRAM yet - I will do this later today, just to confirm what you are seeing.

**Comment by Sebastien-Raguideau** [comment]
I managed to fix my libhts.so and now the new binary works.
I tried a "small" example 100000 bed queries on 1000 contigs, it is too small to incur any memory issue on both regular and fixed mosdepth. Though regarding speed the fixed version is as slow as the regular one, around 30 min when bedtool takes 15 min.
I linked a subsampled cram  + the full bed file :
https://seb.s3.climb.ac.uk/Sharing/sub.cram
https://seb.s3.climb.ac.uk/Sharing/ORFs.bed
will there be any issue if the bed file ask for contigs not in the cram file?
I may find time today to check on the cram file memory issue, but it should take a day to run even without memory overflow.

**Comment by johnomics** [comment]
Can confirm the full CRAM and a 1% subsampled CRAM are fine, 1% subsampled BAM still has the same problem:

![screenshot 2019-02-05 13 27 13](https://user-images.githubusercontent.com/348469/52276457-2feeab00-294a-11e9-8c82-e26d2c68bce0.png)

Free memory plots; full CRAM on left, full BAM on right.

![screenshot 2019-02-05 13 26 16](https://user-images.githubusercontent.com/348469/52276459-32e99b80-294a-11e9-88f8-1534b2a9127e.png)

1% BAM on left, 1% CRAM on right.

Even 1% CRAM is 373 MB so I won't upload now, as you have a BAM already - but can do if you need more examples.

Thanks again for looking into this. Got the output I need by leaving off the -t option.

**Comment by Sebastien-Raguideau** [comment]
Cool, that's one issue fixed.
I just checked for both new and old binary, removing -t does not change the speed of execution : still taking at least twice bedtool's.

**Comment by brentp** [comment]
i'll close this issue for the memory. there's another issue tracking the speed with huge numbers of chromosomes for certain parameter combinations here: https://github.com/brentp/mosdepth/pull/56

**Comment by brentp** [comment]
Hi All, this issue is resolved in #229 and the soon to be released v0.3.8. There were a couple things that were O(n^2) with the number of chroms :man_facepalming: which usually doesn't matter, but does with a large number of chromosomes.
